lib/propolis/src/hw/pci/bar.rs

```diff
@@ -121,33 +121,35 @@ impl Bars {
         &mut self,
         bar: BarN,
         val: u32,
-    ) -> Option<(BarDefine, u64, u64)> {
+    ) -> Option<WriteResult> {
         let idx = bar as usize;
         let ent = &mut self.entries[idx];
-        let (def, old, new) = match ent.kind {
+        let (id, def, val_old, val_new) = match ent.kind {
             EntryKind::Empty => return None,
             EntryKind::Pio(size) => {
                 let mask = u32::from(!(size - 1));
                 let old = ent.value;
                 ent.value = u64::from(val & mask);
-                (BarDefine::Pio(size), old, ent.value)
+                (bar, BarDefine::Pio(size), old, ent.value)
             }
             EntryKind::Mmio(size) => {
                 let mask = !(size - 1);
                 let old = ent.value;
                 ent.value = u64::from(val & mask);
-                (BarDefine::Mmio(size), old, ent.value)
+                (bar, BarDefine::Mmio(size), old, ent.value)
             }
             EntryKind::Mmio64(size) => {
                 let old = ent.value;
                 let mask = !(size - 1) as u32;
                 let low = val & mask;
```

Comment:
i really just noticed it while reading this code, but i think we should mask off the low four bits of `val` for `Mmio` and `Mmmio64` (i.e. the low BAR), 'cause it looks to me like if firmware wrote back the bits we offered we'd faithfully place the bar off-by-one for no good reason.